common4j/src/main/com/microsoft/identity/common/java/telemetry/adapter/BrokerTelemetryAdapter.java

```diff
@@ -21,16 +21,93 @@
 // THE SOFTWARE.
 package com.microsoft.identity.common.java.telemetry.adapter;
 
+import com.microsoft.identity.common.java.telemetry.TelemetryEventStrings;
 import com.microsoft.identity.common.java.telemetry.observers.IBrokerTelemetryObserver;
+import com.microsoft.identity.common.java.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 import lombok.NonNull;
 
 /**
  * Adapter to aggregate events for broker telemetry. We make use of the {@link TelemetryAggregationAdapter} which
  * aggregates the events into MATS format.
+ * <p>
+ * We then send out the error events as single events since the {@link TelemetryAggregationAdapter} will pick the last error event
+ * and add combine it with the aggregated event. This means that the event sent by {@link TelemetryAggregationAdapter} only contains
+ * the last error that was dispatched.
  */
 public class BrokerTelemetryAdapter extends TelemetryAggregationAdapter {
     public BrokerTelemetryAdapter(@NonNull IBrokerTelemetryObserver observer) {
         super(observer);
     }
+
+    @Override
+    public void process(@NonNull List<Map<String, String>> rawData) {
+        final Map<String, String> aggregatedMap = aggregateEvent(rawData);
+
+        // Filter out the error events first
+        final List<Map<String, String>> errorEvents = filterErrorEvents(rawData, aggregatedMap);
+
+        // count the number of hits of the errors
+        final List<Map<String, String>> countedErrors = countErrors(errorEvents);
+        // publish the event
+        for (Map<String, String> errorEvent : countedErrors) {
+            getObserver().onReceived(errorEvent);
+        }
+
+        // send out the aggregated event separately. This contains the tag of the error event that was
+        // fired last, hence in case of a failure, we know which error caused it.
```

Comment:
So we only know tag for one error event?

Reply:
Yes that's right. But in line 59
```java
        for (Map<String, String> errorEvent : countedErrors) {
            getObserver().onReceived(errorEvent);
        }
```

We emit error events separately, hence we can query all errors for a particular event by correlation id.